Feature/hidden pages

skuethe opened the pull request:
PR for #47 

[I had to join the `modules` and `fixed` arrays during the `MM.getModules()` loop](https://github.com/edward-shen/MMM-pages/compare/master...skuethe:feature/hiddenPages?expand=1#diff-645007cc976e2e727b7fd2a610276c4cdfe066543644879e5ca09bba340e8ce3R217). Because defining multiple `.withClass` or `.exceptWithClass` seem to not work and that broke view while returning from a hidden page

An open topic for discussion is how to interact with `MMM-page-indicator`.
First thought was to send a non-existing index to it, so that none of the circle indicators would be "active". But that seems like a dirty hack.
Second thought was to force hide it when entering a hidden page (because it does not make sense to show there, right?). But that also feels wrong. What if, for what ever reason, a user wants to still show it on a hidden page. Then forcing it will just cause problems.
My third and currently implemented thought is to just give the user control.
If you want to hide it, just don't put it into the module array of the named hidden page. And if you really want to show it, just add it.

edward-shen commented on MMM-pages.js:
```diff
@@ -185,31 +185,51 @@ Module.register('MMM-pages', {
-          module.show(
+        .exceptWithClass(this.config.hiddenPages[name])
+        .enumerate(module => module.hide(
+          self.config.animationTime / 2,
+          { lockString: self.identifier }
+        ));
+      // Shows, after a small delay, all modules defined in the selected named hidden page.
+      setTimeout(() => {
+        MM.getModules()
+          .withClass(self.config.hiddenPages[name])
+          .enumerate((module) => {
+            module.show(
+              self.config.animationTime / 2,
+              { lockString: self.identifier }
+            );
+          });
+      }, this.config.animationTime / 2);
+    } else {
+      // Hides all modules not on the current page. This hides any module not
+      // meant to be shown.
+      MM.getModules()
+        .exceptWithClass(this.config.fixed.concat(this.config.modules[this.curPage]))
+        .enumerate(module => module.hide(
+          self.config.animationTime / 2,
+          { lockString: self.identifier }
+        ));
+
+      // Shows all modules meant to be on the current page, after a small delay.
+      setTimeout(() => {
+        MM.getModules()
+          .withClass(self.config.fixed.concat(self.config.modules[self.curPage]))
+          .enumerate(module => module.show(
             self.config.animationTime / 2,
             { lockString: self.identifier }
-          );
-        });
-    }, this.config.animationTime / 2);
+          ));
+      }, this.config.animationTime / 2);
+    }
```
Is the only difference between these two if-branches just which set of modules are excluded?

skuethe replied:
Yes, I fixed that in combination with the local variable.

edward-shen commented on MMM-pages.js:
```diff
@@ -244,4 +264,42 @@ Module.register('MMM-pages', {
+   *
+   * @param {boolean} pause the parameter, if you want to pause or resume.
+   */
+  pauseRotation: function (pause) {
```
Not a big fan of this name -- can we change this to be something like `setRotation: function (isRotating)` instead?

skuethe replied:
fixed

edward-shen commented on readme.md:
```diff
@@ -43,24 +48,29 @@ modules: [
+| `modules`           | `[[String...]...]`         | `[]`                     | A 2D String array of what each module should be on which page. Note that all entries must take their class name (e.g. this module's class name is `MMM-pages`, while the default modules may just have `newsfeed`, without the `MMM-` prefix. |
+| `fixed`             | `[String...]`              | `["MMM-page-indicator"]` | Which modules should show up all the time. |
+| `excludes`          | *NA*                       | *NA*                     | **Deprecated**. Use `fixed` instead. |
+| `hiddenPages`       | `{String: [String...]...}` | `{}`                     | An Object defining special `hiddenPages` which are not available on the normal page rotation and only accassible via a notification. Modules defined in `fixed` are ignored and need to be also added if you wish to have them on any hidden page |
```
missing period

edward-shen commented on readme.md:
```diff
@@ -79,6 +89,8 @@ The following is the list of notifications that MMM-pages will handle:
 | `PAUSE_ROTATION`    | *None*          | If MMM-pages is set to rotate, this will pause rotation until a `RESUME_ROTATION` notification is sent. This does nothing if rotation was already paused. |
 | `RESUME_ROTATION`   | *None*          | If MMM-pages was requested to pause rotation, this will resume automatic rotation. This does nothing MMM-pages was not requested to pause. |
 | `HOME_PAGE`         | *None*          | Return to the home page. If no home page is provided, return to the first page instead. |
+| `SHOW_HIDDEN_PAGE`  | `String`        | MMM-pages will switch to the provided hidden page name |
+| `LEAVE_HIDDEN_PAGE` | *None*          | MMM-pages will leave the currently showing hidden page and return to the previous showing page index |
```
missing periods

edward-shen reviewed (changes requested):
Great PR! Changes look good, except for small nitpicking here and there.

I'll do another review of this PR once these first few have been addressed.

Quick question -- what's your current set up to test this? Have you been testing this with rotation enabled?

Thanks again for the PR!

skuethe commented on MMM-pages.js:
```diff
@@ -77,9 +78,11 @@ Module.register('MMM-pages', {
    *
    * @param {string} notification the notification ID
-   * @param {number} payload the page to change to/by
+   * @param {(number|string)} payload the page to change to/by
```
It seems it's optional - I removed them :)

skuethe commented on MMM-pages.js:
```diff
@@ -171,7 +171,7 @@ Module.register('MMM-pages', {
    */
   updatePages: function () {
-    // Update iff there's at least one page.
+    // Update if there's at least one page.
```
Sh*t :D - reverted

skuethe commented on MMM-pages.js:
```diff
@@ -185,31 +185,51 @@ Module.register('MMM-pages', {
    * assumes that there is a discrepancy between the page currently being shown
    * and the page that is meant to be shown.
+   * 
+   * @param {string} [name] - the name of the hiddenPage we want to show. Optional and only used when we want to switch to a hidden page
```
I removed the hyphen but kept the brackets, as they define an optional parameter in JSDoc syntax.
Still need to look at how to force 80 characters limit and will then update

edward-shen replied:
Ah, thanks for letting me know. If that's correct JSDoc I won't press for it then.

skuethe commented on readme.md:
```diff
@@ -33,7 +36,9 @@ To display what page you're on, I'd highly recommend checking out my
 
-To use this module, add it to the modules array in the `config/config.js` file:
+To use this module, add it to the modules array in the `config/config.js` file.  
+Note: module names used in the following example can be fictitious.
```
Wanted to use that first, but then had some struggle as some of them are not fictitious :D 
But yeah, it's still the better way to go. Fixed.

skuethe commented on MMM-pages.js:
```diff
@@ -244,4 +264,42 @@ Module.register('MMM-pages', {
+   * @param {string} name - the name of the hiddenPage we want to show
+   */
+  showHiddenPage: function (name) {
+    if (typeof name !== 'undefined') {
```
fixed

skuethe commented on MMM-pages.js:
```diff
@@ -185,31 +185,51 @@ Module.register('MMM-pages', {
+   * @param {string} [name] - the name of the hiddenPage we want to show. Optional and only used when we want to switch to a hidden page
    */
-  animatePageChange: function () {
+  animatePageChange: function (name) {
```
Do we need to? In most of the cases this is still used for switching to another page index instead of a named page. In that case `targetPageName` does not seem to fit. I feel `animatedPageChange` describes exactly what this functions main task is :)

edward-shen replied:
Oh, I mean the variable `name`, not `animatedPageChange`.

skuethe replied:
forgot to fix this. I'll do it tomorrow.

skuethe commented:
Thanks for the quick review. Fixed most of them.
I tested automatic rotation and stumbled upon two issues:

1. when switching a page by `rotationTime` there was no interaction with `MMM-page-indicator`
2. `rotationDelay` is set default to `10000` but README suggests its `0`

Instead of changing the default `rotationDelay` to `0` I updated the README, because that will otherwise mess up peoples setup if they automatically `git pull`

edward-shen commented:
Heads up, you'll need to rebase as I needed to do an emergency fix for #49.

edward-shen commented on MMM-pages.js:
```diff
@@ -240,8 +244,47 @@ Module.register('MMM-pages', {
+   * @param {boolean} isRotating the parameter, if you want to pause or resume.
+   */
+  setRotation: function (isRotating) {
+    var stateBaseString = (isRotating) ? "paus" : "resum";
```
Use `const`, please.

edward-shen commented:
> when switching a page by rotationTime there was no interaction with MMM-page-indicator

This was also just fixed upstream. Thanks for notifying me.

edward-shen commented:
Oh my bad, I didn't notice you fixed this on your branch as well -- I would have cherry picked your commit if I saw it. Unfortunately I can't change attribution as it's already public :(

skuethe commented:
Thats OK, I'll just have to fight through the rebase :)

skuethe commented:
Okay, I hope that did the trick and everything is smooth after resolving the conflicts during rebase.

edward-shen commented on MMM-pages.js:
```diff
@@ -113,27 +118,24 @@ Module.register('MMM-pages', {
-          Log.log('[Pages]: resuming rotation due to notification');
-          this.resetTimerWithDelay(this.rotationDelay);
-          this.rotationPaused = false;
-        } else {
-          Log.warn('[Pages]: Was asked to resume but rotation was not paused!');
-        }
+        this.setRotation(false);
```
It feels kinda weird that resuming rotation sets rotation to false, and that pausing rotation sets rotation to true... can we invert the logic in `setRotation`?

skuethe replied:
You are totally right. With renaming the function the reverse logic makes more sense.

edward-shen reviewed (commented):
Looks good! Just one last thing to talk about about, and then we can merge it!

edward-shen reviewed (approved):
Looks good to me. I'll merge this in now. Thanks again for working with me to get this feature in!

skuethe commented:
thank you too for the support :)

lxne commented:
Awesome. Great work.
Thank you so much – both of you.
